python/triton/runtime/autotuner.py

```diff
@@ -137,23 +137,35 @@ def _bench(self, *args, config, **meta):
         current = dict(meta, **config.all_kwargs())
         full_nargs = {**self.nargs, **current}
 
+        # hook arguments are passed as positional-only. here we
+        # collect as many as we can from the full_nargs. some
+        # args in the self.arg_names may be missing in full_nargs
+        # when they are injected by a nested decorator (e.g.,
+        # downstream @triton.heuristics)
```

Comment:
Can you describe the problem here a bit more? It seems bad to just silently drop arguments.

Reply:
The `pre_hook` and `post_hook` of `Autotuner` are currently getting positional arguments by design.

Before this PR, in the `_bench` method (e.g., [here](https://github.com/triton-lang/triton/blob/1cf7b1b31cde8c62611e421becd4648c7284d76c/python/triton/runtime/autotuner.py#L143)), we passed the `args` directly to the `pre_hook` and `post_hook`. This is problematic when the kernel arguments that should be available in the hooks (e.g., the ones mentioned in `restore_value` or `reset_to_zero`) are passed as kwargs to the kernel call: they'll be missing in the `args`. The new variation of the `test_restore` added in this PR fails because of this (also there is a rerpo and error in #5082).

In this PR, instead of passing the `args` to the hooks, we collect the positional arguments from the `full_nargs` (which contains args, kwargs, and config) in the order of `self.arg_names`. Now the problem is that, sometimes, a certain argument in `self.arg_names` is provided by an inner `@triton.heuristics`. In this case, that particular argument won't be available in the `full_nargs` and we have to cut the chain of positional arguments. One alternative would be passing `None` to the hooks for those unavailable arguments and continue traversing the `self.arg_names` till the end, without dropping any arguments that we have access to in the `_bench` method. Do you think this would be better?

I believe, the approach in this PR passes at least as many arguments to the hooks as the current approach in the trunk (probably more, due to covering kwargs and config), so this shouldn't break existing code relying on the hooks.

Reply:
Please verify using https://github.com/triton-lang/triton/blob/main/third_party/proton/tutorials/matmul.py